src/sentry/roles/manager.py

```diff
@@ -31,55 +51,155 @@ def has_scope(self, scope: str) -> bool:
         return scope in self.scopes
 
 
-class RoleManager:
-    def __init__(
-        self,
-        config: Iterable[Mapping[str, str]],
-        default: Optional[str] = None,
-    ) -> None:
-        role_list = []
-        self._roles: Dict[str, Role] = OrderedDict()
-        for idx, role_cfg in enumerate(config):
-            role = Role(idx, **role_cfg)
-            role_list.append(role)
-            self._roles[role.id] = role
+@dataclass(frozen=True, eq=True)
+class OrganizationRole(Role):
+    is_global: bool = False
+
+    def can_manage(self, other: OrganizationRole) -> bool:
+        return self.priority >= other.priority
+
+    def get_minimum_team_role(self) -> TeamRole:
+        """Return the minimum team role for this organization role.
+
+        A member with this organization role automatically receives the minimum role
+        when joining a new team, and can't be demoted below that team role for as
+        long as they hold the organization role.
+        """
+        return self.parent.get_minimum_team_role(self.id)
+
+    def can_manage_team_role(self, other: TeamRole) -> bool:
+        return self.get_minimum_team_role().can_manage(other)
 
-        self._choices = tuple((r.id, r.name) for r in role_list)
 
-        if default:
-            self._default = self._roles[default]
-        else:
-            self._default = role_list[0]
+@dataclass(frozen=True, eq=True)
+class TeamRole(Role):
+    is_minimum_role_for: str | None = None
 
-        self._top_dog = role_list[-1]
+    def can_manage(self, other: TeamRole) -> bool:
+        return self.priority >= other.priority
```

Comment:
Not sure if this is possible with python typing, but can we have this on the base class and have the definition something to the effect of
```
    def can_manage(self, other: type(self)) -> bool:
        return self.priority >= other.priority
```
?

Reply:
👍 Turns out https://github.com/getsentry/sentry/pull/33387/commits/f85f4c156dddf7e929b6e7b7220f9015f9910f30 does the trick.

Reply:
Nice!